### PR TITLE
Update rust-toolchain to nightly of 2023-07-30

### PR DESCRIPTION
**Pull request by lschuermann**
### Pull Request Overview

This version and the time to update is chosen somewhat arbitrarily, as libtock-rs fails to build elf2tab on its current Rust toolchain (2022-06-10). That triggered me to update them in tandem to a recent Rust nightly.

Fixes some new clippy warnings / errors, such as:

```
error: incorrect implementation of `clone` on a `Copy` type
  --> kernel/src/utilities/static_ref.rs:44:29
   |
44 |       fn clone(&self) -> Self {
   |  _____________________________^
45 | |         StaticRef { ptr: self.ptr }
46 | |     }
   | |_____^ help: change this to: `{ *self }`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#incorrect_clone_impl_on_copy_type
   = note: `#[deny(clippy::incorrect_clone_impl_on_copy_type)]` on by default
```

and

```
error: called `map(<f>).unwrap_or(<a>)` on an `Option` value. This can be done more directly by calling `map_or(<a>, <f>)` instead
   --> capsules/extra/src/can.rs:293:17
    |
293 | /                 self.can_rx
294 | |                     .take()
295 | |                     .map(|dest_buffer| {
296 | |                         self.processes
...   |
325 | |                     })
326 | |                     .unwrap_or(CommandReturn::failure(ErrorCode::NOMEM))
    | |________________________________________________________________________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_unwrap_or
    = note: requested on the command line with `-D clippy::map-unwrap-or`
help: use `map_or(<a>, <f>)` instead
    |
295 ~                     .map_or(CommandReturn::failure(ErrorCode::NOMEM), |dest_buffer| {
296 |                         self.processes
  ...
324 |                             .unwrap_or_else(|err| err.into())
325 ~                     })
    |
```

### Testing Strategy

This pull request was tested by CI.

### TODO or Help Wanted

~This introduces a warning in the license-checker stage of the `make prepush` command:~
Figured it out, updated `workspace.resolver = "2"` in the `tools/Cargo.toml` workspace.

### Documentation Updated

- [x] ~Updated the relevant files in `/docs`, or~ no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

**Review by alistair23 — APPROVED** [review]

**Review by bradjc — APPROVED** [review]

File: capsules/extra/src/can.rs [hunk 1]
```diff
@@ -292,7 +292,7 @@ impl<'a, Can: can::Can> SyscallDriver for CanCapsule<'a, Can> {
             7 => {
                 self.can_rx
                     .take()
-                    .map(|dest_buffer| {
+                    .map_or(CommandReturn::failure(ErrorCode::NOMEM), |dest_buffer| {
```
Comment (ppannuto) [inline capsules/extra/src/can.rs:295]: Not that I think there's anything wrong with this, but what triggered this change?

**Review by ppannuto — APPROVED** [review]

File: capsules/extra/src/can.rs (hunk 1 above)
Comment (lschuermann) [reply capsules/extra/src/can.rs:295]: Clippy, trying to make our code more idiomatic. This was a hard-error in CI. I'm not neither happy about nor opposed to it complaining here.

**Comment by lschuermann** [comment]
Marking as blocked until I figure the cargo warning out.

**Comment by lschuermann** [comment]
Figured out the `make prepush` warning, turns out we maintain a second Cargo workspace in `tools/`. After updating the `resolver = "2"` in there this no longer emits warnings. Also rebased & resolved conflicts, this should be good to go now!

**Review by bradjc — APPROVED** [review]

**Review by jrvanwhy — APPROVED** [review]

**Review by ppannuto — APPROVED** [review]

**Review by alistair23 — APPROVED** [review]

**Pull request merged**